extensions/tokenizers/rust/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+    _: JObject,
+    handle: jlong,
+    inputs: jobjectArray,
+    add_special_tokens: jboolean,
+) -> jlongArray {
+    let tokenizer = cast_handle::<Tokenizer>(handle);
+    let len = env.get_array_length(inputs).unwrap();
+    let mut array: Vec<String> = Vec::new();
+    for i in 0..len {
+        let item = env.get_object_array_element(inputs, i).unwrap().into();
+        let value: String = env
+            .get_string(item)
+            .expect("Couldn't get java string!")
+            .into();
+        array.push(value);
+    }
+
+    let encodings = tokenizer
+        .encode_batch(array, add_special_tokens == JNI_TRUE)
+        .unwrap();
+    let handles = encodings
+        .into_iter()
+        .map(|c| to_handle(c))
+        .collect::<Vec<_>>();
+
+    let size = handles.len() as jsize;
+    let ret = env.new_long_array(size).unwrap();
+    env.set_long_array_region(ret, 0, &handles).unwrap();
+    ret
+}
+
+#[no_mangle]
+pub extern "system" fn Java_ai_djl_huggingface_tokenizers_jni_TokenizersLibrary_deleteEncoding(
+    _env: JNIEnv,
+    _: JObject,
+    handle: jlong,
+) {
+    drop_handle::<Encoding>(handle);
+}
+
+#[no_mangle]
+pub extern "system" fn Java_ai_djl_huggingface_tokenizers_jni_TokenizersLibrary_getTokenIds(
+    env: JNIEnv,
+    _: JObject,
+    handle: jlong,
+) -> jlongArray {
+    let encoding = cast_handle::<Encoding>(handle);
+    let ids = encoding.get_ids();
+    let len = ids.len() as jsize;
+    let mut long_ids: Vec<jlong> = Vec::new();
+    long_ids.reserve(len as usize);
+    for i in ids {
+        long_ids.push(*i as jlong)
+    }
+
+    let array: jlongArray = env.new_long_array(len).unwrap();
+    env.set_long_array_region(array, 0, &long_ids).unwrap();
+    array
+}
+
+#[no_mangle]
+pub extern "system" fn Java_ai_djl_huggingface_tokenizers_jni_TokenizersLibrary_getTypeIds(
+    env: JNIEnv,
+    _: JObject,
+    handle: jlong,
+) -> jlongArray {
+    let encoding = cast_handle::<Encoding>(handle);
+    let type_ids = encoding.get_type_ids();
+    let len = type_ids.len() as jsize;
+    let mut long_ids: Vec<jlong> = Vec::new();
+    for i in type_ids {
+        long_ids.push(*i as jlong)
+    }
+
+    let array: jlongArray = env.new_long_array(len).unwrap();
+    env.set_long_array_region(array, 0, &long_ids).unwrap();
+    array
+}
+
+#[no_mangle]
+pub extern "system" fn Java_ai_djl_huggingface_tokenizers_jni_TokenizersLibrary_getWordIds(
+    env: JNIEnv,
+    _: JObject,
+    handle: jlong,
+) -> jlongArray {
+    let encoding = cast_handle::<Encoding>(handle);
+    let word_ids = encoding.get_word_ids();
+    let len = word_ids.len() as jsize;
+    let mut long_ids: Vec<jlong> = Vec::new();
+    for i in word_ids {
+        if let Some(word_id) = i {
+            long_ids.push(*word_id as jlong)
+        } else {
+            long_ids.push(-1)
+        }
+    }
+
+    let array: jlongArray = env.new_long_array(len).unwrap();
+    env.set_long_array_region(array, 0, &long_ids).unwrap();
```

Comment:
I don't know JNI well: does this mean that the VM becomes responsible for GCing the jlongArray? (which would be memory safe)

Reply:
Yes, JVM will collect the return long[]